confdgnmi/README.md

```diff
@@ -0,0 +1,283 @@
+
+Options:
+  -h, --help            show this help message and exit
+  -o OPERATION, --oper=OPERATION
+                        gNMI operation [capabilities, set, get, subscribe]
+  --logging=LOGGING     Logging level [error, warning, info, debug]
+  --prefix=PREFIX       'prefix' path for set, get and subscribe operation
+                        (empty by default)
+  -p PATHS, --path=PATHS
+                        'path' for get, set and subscribe operation, can be
+                        repeated (empty by default)
+  -v VALS, --val=VALS   'value' for set operation, can be repeated (empty by
+                        default)
+```
+      
+Examples:
+
+`./src/confd_gnmi_client.py -o capabilities`
+`./src/confd_gnmi_client.py -o  get --prefix /interfaces --path interface[name=if_82]/name --path interface[name=if_82]/type`
+`./src/confd_gnmi_client.py -o set  --prefix /interfaces --path interface[name=if_82]/type --val fastEther`
+`./src/confd_gnmi_client.py -o subscribe --prefix /interfaces --path interface[name=if_82]/name --path interface[name=if_82]/type`
+
+NOTE: Other parameters (username, password, subscription types are currently hardcoded, TODO)
+    
+## Limitations and TODOs of current implementation
+
+* `encoding` - used only `BYTES`
+* `Get`, `Set` and `Subscribe` works only on `leaf` elements
+* `Subscribe` - only `ONCE` and `POLL` is implemented so far
+    * `updates_only` not supported
+    * `heartbeat_interval` not supported
+    * `sync_response` not generated
+* all values `TypedValues` are used as strings (`string_val`)
+* gNMI Path is converted to XPath or formatted path with simple string operations 
+  no datamodel knowledge used) 
+  
+## Investigations/TODOs
+
+TODO converting XPath to ConfD formatted path/keypath and back.
+(Work with `cs_node` for API adapter, for Netconf adapter ??)
+
+## Analysis
+
+We use (part) of `ietf-interfaces.yang` as sample datamodel.
+
+### gNMI interface 
+
+#### Capabilities
+
+`rpc Capabilities(CapabilityRequest) returns (CapabilityResponse);`
+
+`Capability` returns list of `supported_models` and list of `encodings`.
+It can also return list of gNMI extensions.  
+
+Each model has following attributes: 
+ 
+`name` (string), `organization` (string), `version`
+
+##### Implementation 
+
+Capability information can be fetched from `"/ncm:netconf-state/ncm:capabilities/ncm:capability"`  
+datamodel, found in the `ietf-netconf-monitoring.yang`.  
+
+Another option can be to get it with `confd_get_nslist` API call:
+
+Call  `confd_get_nslist`, array of following is returned:
+```
+struct confd_nsinfo {
+    const char *uri;
+    const char *prefix;
+    u_int32_t hash;
+    const char *revision;
+    const char *module;
+}; 
+```
+
+`name` -> `module`
+`organization` -> TODO (cannot get with ConfD API, return empty string)
+`version` -> `revision` (return revision string -date ?)
+
+TODO: Pass yang file content as extension? HOW to get yang files from ConfD?
```

Comment:
Perhaps using RESTCONF  "GET /restconf/tailf/modules/dhcp/2019-02-14"

Reply:
RESTCONF can be added in next milestones